src/test/ui/error-codes/E0121.rs

```diff
@@ -1,6 +1,10 @@
 fn foo() -> _ { 5 } //~ ERROR E0121
 
 static BAR: _ = "test"; //~ ERROR E0121
+//~^ ERROR E0121
```

Comment:
There's a duplicate error now, but as this is a pre-existing issue (https://github.com/rust-lang/rust/issues/77428), and this fixes an ICE that could reasonably be encountered by users, I thought a quick fix was best.

Reply:
Actually, this triggers quite a few more additional error messages in other places; it'd be nicer to fix this now.

Reply:
@estebank: I thought we deduplicated diagnostics? Do you know the reason it's failing in this case?

Reply:
We don't deduplicate automatically at the error level, we do it on a case by case basis ahead of time. For `_` E0121 errors it was a bit ad-hoc (introduced in https://github.com/rust-lang/rust/pull/69148).

Reply:
> I *think* you can get away by looking at the type we're evaluating...

What do you mean by this?